### DISCUSSION TITLE
Plans for supporting Rust plugins?

### DISCUSSION THREAD

**Discussion by phoenix-ru** [vercel/turborepo#2539] [Ideas]
I've heard a lot of mentions that Turbopack aims at becoming an ecosystem, a "successor to Webpack". And thus I am excited about the possibilities for creating non-JS based tooling.

However, based on what esbuild offers in terms of writing native plugins (tl;dr you have to compile your plugin as a binary which calls esbuild) and the fact that you absolutely can't force JS devs to compile their own bundler, what are the plans for plugins?

Just a glimpse of what could be done:
1. Include the most popular plugins into the bundler itself (basically what Webpack did),
2. Add support for JS plugins which in turn can use NAPI or WASM,
3. Allow WASM plugins in the bundler itself, or
4. Provide composable (hookable) API so that build chains (like Next, Nuxt, etc.) would provide framework-specific binaries.

And while 1 and 2 raise no questions, as a maintainer of various huge (3-12K modules) org projects, I often find that "out-of-the-box" experience never works. That is part of the reason we are stuck with Webpack, because no other tool offers such flexibility, especially when it comes to patching some quirks of a framework (Nuxt).

Having tried esbuild, I can pretty much say that simply plugging esbuild-loader to Webpack is a much more pleasant experience than writing plugins for it.

I understand that writing native/wasm plugins is not as straightforward as a js plugin, but when performance is concerned, I personally am ready to invest some time to shave off minutes of developers time every `npm run dev` or every CI/CD run.

**Comment by kwonoj** [comment]
There are internal discussions about the extensibility of turbopack definitely, but so far none of them are concrete enough to be shared publicly yet. SWC already has its experimental plugin support in next.js, which we are carefully evaluating in relation to improvement of turbopack in the future.
